nbgrader/apps/zipcollectapp.py

```diff
@@ -0,0 +1,486 @@
+                }, ...
+            }
+        """
+        self.log.info("Start collecting files...")
+        released_path = self._format_path(
+            self.release_directory, '.', self.assignment_id)
+        released_notebooks = find_all_notebooks(released_path)
+        if not released_notebooks:
+            self.log.warn(
+                "No release notebooks found for assignment {}, did you forget "
+                "to run 'nbgrader assign'?".format(self.assignment_id)
+            )
+
+        data = dict()
+        invalid_files = 0
+        processed_files = 0
+        for _file in src_files:
+            self.log.info("Processing file: {}".format(_file))
+            info = self.collector_plugin_inst.collect(_file)
+            if not info or info is None:
+                self.log.warn("Skipped. No match information provided.")
+                invalid_files += 1
+                continue
+
+            for key in ['student_id', 'file_id']:
+                if key not in info.keys():
+                    self.fail(
+                        "Expected processor {} to provide the {} from the "
+                        "submission file name."
+                        "".format(self.collector_plugin.__name__, key)
+                    )
+                if not isinstance(info[key], six.string_types):
+                    self.fail(
+                        "Expected processor {} to provide a string for {} from "
+                        "the submission file name."
+                        "".format(self.collector_plugin.__name__, key)
+                    )
+
+            root, ext = os.path.splitext(_file)
+            file_id = os.path.splitext(os.path.basename(info['file_id']))[0]
+            submission = '{}{}'.format(file_id, ext)
+            if ext in ['.ipynb'] and submission not in released_notebooks:
+                if self.strict:
+                    self.log.warn("Skipped. Invalid notebook name.")
+                    invalid_files += 1
+                    continue
+                self.log.warn("Invalid notebook name.")
+
+            if self._find_student(info) is None:
+                self.log.warn(
+                    "Skipped. Student {} not found in gradebook."
+                    "".format(info['student_id'])
+                )
+                invalid_files += 1
+                continue
+
+            submitted_path = self._format_path(
+                self.submitted_directory, info['student_id'], self.assignment_id)
+            dest_path = os.path.join(submitted_path, submission)
+
+            timestamp = self.get_timestamp()
+            if 'timestamp' in info.keys():
+                timestamp = info['timestamp'] or timestamp
+                if isinstance(timestamp, six.string_types):
+                    try:
+                        timestamp = parse_utc(timestamp)
+                    except ValueError:
+                        self.log.warn(
```

Comment:
I am not sure if this is the right behavior. I would be inclined to say if there is no timestamp associated with a file, then it doesn't get a timestamp at all.

In particular, if the individual files don't have timestamps associated with them but there is one file in the submission, e.g. `timestamp.txt` that contains a timestamp, then what will happen is all the files get the current time. Then below (line 440), the max of all the timestamps is taken. This means that even if a user writes a custom plugin to parse the timestamp out of `timestamp.txt`, the default timestamp will always be used since it is guaranteed to be newer.